### ISSUE TITLE
Kafka Binding v1.10.4: kafka: client has run out of available brokers to talk to

### ISSUE THREAD

**Issue by Andemki** [dapr/components-contrib#2755]
## What version of Dapr?

1.10.4

## Expected Behavior

Dapr and Azure Event Hub integration using Kafka binding work without errors

## Actual Behavior

The same issue as I had for version 1.8: https://github.com/dapr/dapr/issues/5342

I've been using bindings.kafka component for recieving messages from Azure Event Hub in my application. Everything work properly in Dapr versions 1.8.6. But when I updated dapr components to version 1.10.4, the sidecar container (daprd) of my application's pod failed to start and I received the following error message:

> level=fatal msg="process component my-eventhub-receiver error: [INIT_COMPONENT_FAILURE]: initialization error occurred for my-eventhub-receiver (bindings.kafka/v1): kafka: client has run out of available brokers to talk to: read tcp 10.240.0.177:39836->20.40.127.97:9093: read: connection reset by peer

Yaml output of my bindings.kafka component:

```
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  annotations:
    meta.helm.sh/release-name: my-eventhublistener
    meta.helm.sh/release-namespace: mynamespace
  creationTimestamp: "2023-04-05T11:55:46Z"
  generation: 1
  labels:
    app.kubernetes.io/managed-by: Helm
  name: my-eventhub-receiver
  namespace: mynamespace
  resourceVersion: "217779032"
  uid: 28628338-0e88-4fe3-886a-0a3bb8dcce90
spec:
  metadata:
  - name: brokers
    value: my-eh-namespace.servicebus.windows.net:9093
  - name: topics
    value: my-eh-data
  - name: consumerGroup
    value: my-eventhublistener
  - name: authRequired
    value: "true"
  - name: saslUsername
    value: $ConnectionString
  - name: saslPassword
    secretKeyRef:
      key: connectionString
      name: my-eventhublistener-eventhub
  type: bindings.kafka
  version: v1
```

## Steps to Reproduce the Problem

1. Create a kafka.binding component to receive messages from Event Hub in your application
2. Deploy your application and the component to the Kubernetes
3. Check the daprd component is not in Running state.

**Comment by ItalyPaleAle** [comment]
@mukundansundar @DeepanshuA @skyao you're the Kafka experts, could you please take a look at this?

**Comment by skyao** [comment]
> kafka: client has run out of available brokers to talk to: read tcp 10.240.0.177:39836->20.40.127.97:9093: read: connection reset by peer

It seems that the kafka broker rejuected the connection from daprd, so please help to check:

- if the dns is correct: my-eh-namespace.servicebus.windows.net  to 20.40.127.97 
- if the username / password are correct: since we don't have the information to expand these parameters to get the final username and password

```yaml
  - name: saslUsername
    value: $ConnectionString
  - name: saslPassword
    secretKeyRef:
      key: connectionString
      name: my-eventhublistener-eventhub
```

**Comment by Andemki** [comment]
> > kafka: client has run out of available brokers to talk to: read tcp 10.240.0.177:39836->20.40.127.97:9093: read: connection reset by peer
> 
> It seems that the kafka broker rejuected the connection from daprd, so please help to check:
> 
> * if the dns is correct: my-eh-namespace.servicebus.windows.net  to 20.40.127.97
> * if the username / password are correct: since we don't have the information to expand these parameters to get the final username and password
> 
> ```yaml
>   - name: saslUsername
>     value: $ConnectionString
>   - name: saslPassword
>     secretKeyRef:
>       key: connectionString
>       name: my-eventhublistener-eventhub
> ```
@skyao Yes, dns and username / password are correct as well as other configuration should be correct because as I mentioned in the description, it works properly in Dapr v.1.8.6. But it doesn't work in v.1.10.4. So when I downgrade the Dapr's version back to 1.8.6, it starts to work properly again. So it seems that something has been changed in new versions of Dapr.
When I updated Dapr from version 1.6 to version 1.8 I had the similar issue with bindings.kafka component and you helped me with it here: https://github.com/dapr/dapr/issues/5342

**Comment by kalyankdas** [comment]
I have the same issue as well and already approached DAPR team for support on this. This is prohibiting us to use DAPR for our use case.

I spent a lot of time to make kafka binding work with azure event hub kafka surface with different variation of params. I am getting exact same error as well. 

BTW, when I connect directly with kafka- python , I am able to connect same kafka surface on same event hub, with same settings.  I can connect and publish as well  without any issue. The main.py and requirements is attached. So no issues in the network or firewall.

My binding is below,
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: kafka-binding
spec:
  #type: pubsub.kafka
  type: bindings.kafka
  version: v1
  metadata:
  - name: brokers # Required. Kafka broker connection setting
    value: "<redacted>:9093"
#  - name: consumerGroup # Optional. Used for input bindings.
#    value: "group1"
  - name: topics
    value: "domainevents-airline"
  - name: authRequired
    value: "true"
  - name: clientID # Optional. Used as client tracing ID by Kafka brokers.
    value: "my-dapr-app-id"
  - name: authType # Required.
    value: "password"
  - name: saslUsername # Required if authType is `password`.
    value: "$ConnectionString"
  - name: saslPassword # Required if authType is `password`.
    value: <redacted>
  - name: saslMechanism
    value: "Plain"
  - name: skipVerify
    value: "true"

**Comment by Andemki** [comment]
@kalyankdas As a quick solution you can try to downgrade Dapr to version 1.8.6 and check if the kafka.binding works in this version. It works in my case in version 1.8.6. So I can't update Dapr's version because of this issue and not sure about higher versions before 1.10.4.

**Comment by Andemki** [comment]
@skyao @ItalyPaleAle Could you please tell me if you or any other members of the Dapr team are currently working on this item?

**Comment by DeepanshuA** [comment]
@Andemki @kalyankdas While I am trying to setup to try replicate it locally, could you please let me know if you can try the same with dapr version 1.10.2 and if issue occur with that version. I am just trying to zero down on the change which might have introduced it.

**Comment by Andemki** [comment]
@DeepanshuA Hello, yes I can confirm that there is the same issue with Dapr v.1.10.2

**Comment by DeepanshuA** [comment]
Update:
I am able to replicate the issue on my local - I used eventhubs with kafka.
What I have observed is that this issue is occurring with sarama 1.38.0 (inclusive) onwards. If with master branch checked-out, I downgrade sarama lib to 1.37.2, the issue doesn't occur.
I tried to look into comparison between 1.37.2 and 1.38.0, but don't seem to have a clear cut answer yet (https://github.com/Shopify/sarama/compare/v1.37.2...v1.38.0).

It fails at this place: https://github.com/Shopify/sarama/blob/9127f1c432c08b9a766fdb0931bfd0da80dddde0/broker.go#L1133

I have also logged an issue with sarama https://github.com/Shopify/sarama/issues/2470

**Comment by yaron2** [comment]
@DeepanshuA does this happen with non-Eventhubs Kafka as well?

**Comment by yaron2** [comment]
Ok I see the original issue is with Event Hubs and Kafka specifically. Downgrading a package dependency should be a last resort.

**Comment by Andemki** [comment]
@berndverst @yaron2 I am not sure why this bug item was closed, but we have the same error message in Dapr v.1.10.6.

**Comment by Andemki** [comment]
@DeepanshuA I tried to install the latest Dapr release v1.10.7 and I don't see any error messages in daprd containter but it seems that the kafka.binding still doesn't work. The application which uses this binding doesn't receive messages from EventHub. I tried to install different versions of Dapr and the latest version where it works is 1.8.7. So it doesn't work starting from version 1.9.0. I didn't notice any changes in [Kafka binding spec](https://docs.dapr.io/reference/components-reference/supported-bindings/kafka/). So could you please tell me what can be a reason? Are there any configuration changes for kafka.binding starting from v. 1.9.0? Shall I create a separate item for this issue?

**Comment by DeepanshuA** [comment]
@Andemki - just saw your comment.
Well, the issue reported here seemed to have got fixed in 1.10.7 - as I received confirmation from @kalyankdas and also, could verify that on my setup.
No, there need not be any such configuration change, as far as I remember.

So, are you not even able to publish message to eventhub binding with kafka?

**Comment by berndverst** [comment]
The reported issue is fixed in 1.10.7 and in Dapr 1.11.

@Andemki can you please open a new issue and provide explicit steps to reproduce whatever you are experiencing against Dapr 1.10.7?

**Comment by Andemki** [comment]
@berndverst @DeepanshuA I've created a new issue: https://github.com/dapr/components-contrib/issues/2874
